scrapper.py

```diff
@@ -23,75 +36,310 @@ class IncorrectNumberOfArticlesError(Exception):
-        pass
+    headers = {'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 '
+                             '(KHTML, like Gecko) Chrome/88.0.4324.182 Safari/537.36'}
+
+    def __init__(self,
+                 seed_urls: list,
+                 max_articles: int = None,
+                 max_articles_per_seed: int = None):
+        self.seed_urls = seed_urls  # to parse
+        self.max_articles = max_articles
+        self.max_articles_per_seed = max_articles_per_seed
+        self.seen_urls = set()
+        self.urls = []  # articles urls
 
     @staticmethod
-    def _extract_url(article_bs):
-        pass
+    def _extract_url(article_bs, article_urls, seen_urls):
+        new_article_urls = set()
+        for link in article_bs.find_all('a'):
+            if href := link.get('href'):
+                if res := re.findall(r'https://www.zvezdaaltaya.ru/'
+                                     r'\d{4}/\d{2}/.+/$', href):
+                    if (res[0] not in article_urls
+                            and res[0] not in new_article_urls
+                            and res[0] not in seen_urls):
+                        new_article_urls.add(res[0])
+        if new_article_urls:
+            with open(constants.ARTICLE_URLS, 'a', encoding='utf-8') as file:
+                file.write('\n'.join(new_article_urls) + '\n')
+        return new_article_urls
 
     def find_articles(self):
         """
         Finds articles
         """
-        pass
+        article_id = 1
+        while self.seed_urls:
+            seed_url = self.seed_urls.pop()
+            try:
+                self.process_page(seed_url, article_id)
+            except BadStatusCode:
+                continue
+            else:
+                if len(self.urls) > self.max_articles:
+                    self.urls = self.urls[:self.max_articles]
+                    break
+                article_id += 1
+                if article_id >= self.max_articles:
+                    break
 
-    def get_search_urls(self):
+    @staticmethod
+    def get_page(url):
+        """
+        Returns requests page
+        """
+        response = requests.get(url, headers=Crawler.headers)
+        response.encoding = 'utf-8'
+        if response.status_code == 200:
+            page = response.text
+        else:
+            raise BadStatusCode
+        sleep(randint(3, 10))
+        return page
+
+    def process_page(self, url, article_id):
+        """
+        Processes page and get artcle urls
+        """
+        page = self.get_page(url)
+        with open(os.path.join(constants.PAGES_ROOT, f'{article_id}.html'),
+                  'w', encoding='utf-8') as file:
+            file.write(page)
+        soup = BeautifulSoup(page, 'html.parser')
+        article_urls = self._extract_url(soup, self.urls, self.seen_urls)
+        self.urls.extend(article_urls)
+        return soup
+
+
+class CrawlerRecursive(Crawler):
```

Comment:
it is stated to be recursive, where is the recursion?

Reply:
did it